### PR TITLE
BP-42: new api list ledgers

### PR DESCRIPTION
**Pull request by nicoloboschi**
Descriptions of the changes in this PR:

### Motivation

The new Client API (org.apache.bookkeeper.client.api.BookKeeper) aims to replace obsolete BookKeeper API but some features are not implemented yet, like the functionalities provided by BookKeeperAdmin.
For example, it does not expose a method to list available ledgers, comparable to BookKeeperAdmin#listLedgers().
low.

### Changes

- Ledgers listing
- Simpler access to LedgerMetadata
- Ledger id inside LedgerMetadata

Master Issue: #2422

> ---
> In order to uphold a high standard for quality for code contributions, Apache BookKeeper runs various precommit
> checks for pull requests. A pull request can only be merged when it passes precommit checks.
>
> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
>
> If this PR is a BookKeeper Proposal (BP):
>
> - [ ] Make sure the PR title is formatted like:
>     `<BP-#>: Description of bookkeeper proposal`
>     `e.g. BP-1: 64 bits ledger is support`
> - [ ] Attach the master issue link in the description of this PR.
> - [ ] Attach the google doc link if the BP is written in Google Doc.
>
> Otherwise:
> 
> - [ ] Make sure the PR title is formatted like:
>     `<Issue #>: Description of pull request`
>     `e.g. Issue 123: Description ...`
> - [ ] Make sure tests pass via `mvn clean apache-rat:check install spotbugs:check`.
> - [ ] Replace `<Issue #>` in the title with the actual Issue number.
> 
> ---

### CODE REVIEW

**Comment by nicoloboschi** [comment]
@eolivelli @sijie can you have a look at it?
Also I'm not able to add BP label to the PR

File: site/bps/BP-42-new-api-list-ledgers.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    // new interface
+    interface LedgersIterator {
+
+        boolean hasNext() throws IOException;
+
+        long next() throws IOException;
+    }
+
+    // new interface
+    interface ListLedgersResult extends AutoCloseable {
```
Comment (eolivelli) [inline site/bps/BP-42-new-api-list-ledgers.md:36]: Can you please add a paragraph that explains why we are implementing AutoCloseable?

**Review by eolivelli — CHANGES_REQUESTED** [review]
Great work.
I left a few minor comments

**Review by sijie — APPROVED** [review]

File: site/bps/BP-42-new-api-list-ledgers.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+- Handle metadata driver errors, since the IOException is directly thrown up to caller, allowing user to handle network errors in a more suitable way.
+- Leave the possibility to restrict/filter returned ledgers in future, without API breaking changes   
+
+The implementation will be the same used in BookKeeperAdmin, iterating over `LedgerRangeIterator`, which already handles ledgers search properly.
+
+#### Simpler access to LedgerMetadata
+The implementation will use LedgerManager to retrieve metadata for a specified ledgerId.  
+
+#### Ledger id inside LedgerMetadata
+Each time a LedgerMetadata instance is created, the ledgerId is known, so it is trivial to set it in the instance.
```
[the remark this answers is not in the document]
Comment (nicoloboschi) [reply site/bps/BP-42-new-api-list-ledgers.md:81]: I think yes, when we read metadata from ZK we add the ledgerId and we don't modify serialization/deserialization

**Comment by nicoloboschi** [comment]
I just have pushed fixes requested from @eolivelli, I will submit a implementation PR in the next days

**Review by eolivelli — APPROVED** [review]
Looks good to me.
Looking forward to the implementation 

Thanks

**Pull request merged**